README.md
## Hydrolysis Reactor - Extract Oxygen From Water!
✔ New extractor takes water from deposits, converting some of it to oxygen <br>
✔ Excess water is usable in the life support grid like any other water <br>
✔ Additionally produces +100% oxygen and power with upgrades <br>


### Description:
New building 'Hydrolysis Reactor' takes 3 water from underground deposits and converts it to oxygen. 2 water which is not converted can be used in the life support grid like any other water. Tourists find the Hydrolysis Reactor to be an amazing feat of Martian science and engineering, making it a good stop for tours.

The reactor requires maintenance of Machine Parts x2 and a water deposit to continue working. It costs the following resources to build:  <br>
    ✔ Concrete x5 <br>
    ✔ Metals x4 <br>
    ✔ Electronics x1 <br>
    ✔ Machine Parts x2 <br>

The Hydrolysis Reactor has three upgrades available.

1) Magnetic Filtering (works exactly the same as a MOXIE) <br>
   ✔ Unlocked by researching 'Magnetic Filtering' <br>
   ✔ +50% oxygen production <br>
   ✔ Costs Polymers x5 <br>


2) Advanced Reactions <br>
   ✔ Unlocked by researching 'Nuclear Fusion' <br>
   ✔ Power Production = 2 * Oxygen Production <br>
   ✔ Costs Electronics x2 and Rare Metals x1 <br>


3) Moisture Farming <br>
   ✔ Unlocked by researching 'Moisture Farming' <br>
   ✔ +50% oxygen <br>
   ✔ Costs Polymers x5 and Metals x2 <br>

This building can be built in domes.

### Options:
    Disable MOXIEs
        On / Off

### Special Thanks:
As requested by *LoneSamurai*, thanks for the mod idea!

## Notes:
If you have any feedback please contact me directly instead of reporting the mod. I'm always happy to take requests as well.<br>
I hang out in [*Discord*](https://discord.gg/kTgYq9UjWx) and on [*Reddit*](https://www.reddit.com/user/Fizzle_Fuze)

Please take a second to rate this mod and leave a comment.<br>
Your feedback helps me produce better quality mods!

Have a good game,<br>
Fizzle Fuze :)

## Info:

### Changelog:
Details on the latest changes as well as changes in future versions are found in [*Info/changelog.md*](Info/changelog.md)

### Contact Me:
I appreciate feedback on my mods and ideas for new ones!

    ✔ Discord:https://discord.gg/kTgYq9UjWx
    ✔ Reddit: /u/Fizzle_Fuze (/r/SurvivingMars)
    ✔ GitHub: https://github.com/orgs/Surviving-Mars-Mods
    ✔ Email: mods@fizzlefuze.com

See [*Info/ContactMe.txt*](Info/ContactMe.txt) for more info on contacting me.

### Other Locations:
My mods are also available on Steam and the Paradox Website:

[Steam Workshop](https://steamcommunity.com/id/fizzle_fuze/myworkshopfiles/?appid=464920) <br>
[Paradox Plaza](https://mods.paradoxplaza.com/games/surviving_mars?search=Fizzle_Fuze&sortBy=best)


### License:
See [*Info/License.md*](Info/license.md) and [*Info/LICENSE*](Info/LICENSE) for copyright and licensing info.


changelog.md
Version 1.2
- B&B no longer required!
- Model is automatically set based on available content (Below & Beyond, Content Pack 3, Base Game)
- Can be built in domes
- Power output changed to double the oxygen output instead of +5
- New upgrade 'Moisture Farming', unlocked by researching 'Moisture Farming', which increases oxygen output by 50% (increasing power output as well)
- Mod option to disable MOXIEs

Version 1.1
- Creation

Code/Code.lua
--Copyright
--[[
*******************************************************************************
Fizzle_Fuze's Surviving Mars Mods
Copyright (c) 2022 Fizzle Fuze Enterprises (mods@fizzlefuze.com)

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU Affero General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Affero General Public License for more details.

    You should have received a copy of the GNU Affero General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.

  If your software can interact with users remotely through a computer
network, you should also make sure that it provides a way for users to
get its source.  For example, if your program is a web application, its
interface could display a "Source" link that leads users to an archive
of the code.  There are many ways you could offer source, and different
solutions will be better for different programs; see section 13 for the
specific requirements.

  You should also get your employer (if you work as a programmer) or school,
if any, to sign a "copyright disclaimer" for the program, if necessary.
For more information on this, and how to apply and follow the GNU AGPL, see
<https://www.gnu.org/licenses/>.
*******************************************************************************
--]]

--mod name
local ModName = "["..CurrentModDef.title.."]"

--translation strings
local Translate = { ID = {}, Text = {} }

Translate.Text['MoxieDisable'] = "Replaced by Hydrolysis Reactor."

--get every string a unique ID
for k, _ in pairs(Translate.Text) do
    Translate.ID[k] = RandomLocId()
    if not Translate.ID[k] then
        Log("ERROR", "Could not find valid translation ID for '", k, "'!")
    end
end

--logging variables
local Debugging = false

--print log messages to console and disk
local function PrintLog()
    local MsgLog = SharedModEnv["Fizzle_FuzeLog"]

    if #MsgLog > 0 then
        --print logged messages to console and file
        for _, Msg in ipairs(MsgLog) do
            print(Msg)
        end
        FlushLogFile()

        --reset
        SharedModEnv["Fizzle_FuzeLog"] = {}
        return
    end
end

--setup cross-mod variables for log if needed
if not SharedModEnv["Fizzle_FuzeLog"] then
    SharedModEnv["Fizzle_FuzeLog"] = { ModName.." INFO: First Fizzle_Fuze mod loading!" }
end

--main logging function
function Fizzle_FuzeLogMessage(...)
    local Sev, Arg = nil, {...}
    local SevType = {"INFO", "DEBUG", "WARNING", "ERROR", "CRITICAL"}

    if #Arg == 0 then
        print(ModName,"/?.lua CRITICAL: No error message!")
        FlushLogFile()
        MsgLog[#MsgLog+1] = ModName.."/?.lua CRITICAL: No error message!"
        SharedModEnv["Fizzle_FuzeLog"] = MsgLog
        return
    end

    for _, ST in ipairs(SevType) do
        if Arg[2] == ST then --2nd arg = severity
            Arg[2] = Arg[2]..": "
            Sev = Arg[2]
            break
        end
    end

    if not Sev then
        Sev = "DEBUG: "
        Arg[2] = "DEBUG: "..Arg[2]
    end

    if (Sev == "DEBUG: " and Debugging == false) or (Sev == "INFO: " and Info == false) then
        return
    end

    local MsgLog = SharedModEnv["Fizzle_FuzeLog"]
    local Msg = ModName.."/"..Arg[1]..".lua "
    for i = 2, #Arg do
        Msg = Msg..tostring(Arg[i])
    end
    MsgLog[#MsgLog+1] = Msg
    SharedModEnv["Fizzle_FuzeLog"] = MsgLog

    if (Debugging == true or Info == true) and Sev == "WARNING" or Sev == "ERROR" or Sev == "CRITICAL" then
        PrintLog()
    end
end

--wrapper logging function for this file
local function Log(...)
    Fizzle_FuzeLogMessage("Code", ...)
end

--unlock upgrade
local function UnlockUpgrade(Upgrade)
    if UIColony then
        UIColony:UnlockUpgrade(Upgrade)
    end
end

--setup options and config
local function UpdateOptions()
    local DisableMOXIE = CurrentModOptions:GetProperty("DisableMOXIE")
    local Entity
    Log("BEGIN UpdateOptions()")

    if IsDlcAccessible("picard") then
        Entity = "Electrolyzer"
    elseif IsDlcAccessible("contentpack3") then
        Entity = WaterExtractorCP3
    end

    if Entity and ClassTemplates.Building.FFHydrolysisReactor then
        Log("Updating Class Template...")
        ClassTemplates.Building.FFHydrolysisReactor.entity = Entity
    end

    if UICity then
        if UICity.labels.FFHydrolysisReactor then
            Log("Updating entity of existing reactors...")
            for _, HR in ipairs(UICity.labels.FFHydrolysisReactor) do
                HR:ChangeEntity(Entity)
            end
        end

        if IsTechResearched("NuclearFusion") then
            UnlockUpgrade("HydrolysisReactor_AdvancedReactions")
        end
        if IsTechResearched("MoistureFarming") then
            UnlockUpgrade("HydrolysisReactor_MoistureFarming")
        end
    end

    if DisableMOXIE then
        LockBuilding("MOXIE", "disable", T(Translate.ID['MoxieDisable'], Translate.Text['MoxieDisable']))
    else
        RemoveBuildingLock("MOXIE")
    end

    Log("FINISH UpdateOptions()")
end

--event handling
function OnMsg.NewHour()
    if Debugging == true then
        PrintLog()
    end
end

--event handling
function OnMsg.NewDay()
    PrintLog()
end

--event handling
function OnMsg.ApplyModOptions(id)
    if id == CurrentModId then
        UpdateOptions()
    end
end

--event handling
OnMsg.ModsReloaded = UpdateOptions
OnMsg.CityStart = UpdateOptions
OnMsg.LoadGame = UpdateOptions

--event handling
function OnMsg.TechResearched(tech_id, _, first_time)
    Log("Tech, FirstTime = ", tech_id, ", ", first_time)
    if tech_id == "NuclearFusion" and first_time then
        UnlockUpgrade("HydrolysisReactor_AdvancedReactions")
    end
    if tech_id == "MoistureFarming" and first_time then
        UnlockUpgrade("HydrolysisReactor_MoistureFarming")
    end
end

Code/HydrolysisReactor.lua
--Copyright
--[[
*******************************************************************************
Fizzle_Fuze's Surviving Mars Mods
Copyright (c) 2022 Fizzle Fuze Enterprises (mods@fizzlefuze.com)

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU Affero General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Affero General Public License for more details.

    You should have received a copy of the GNU Affero General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.

  If your software can interact with users remotely through a computer
network, you should also make sure that it provides a way for users to
get its source.  For example, if your program is a web application, its
interface could display a "Source" link that leads users to an archive
of the code.  There are many ways you could offer source, and different
solutions will be better for different programs; see section 13 for the
specific requirements.

  You should also get your employer (if you work as a programmer) or school,
if any, to sign a "copyright disclaimer" for the program, if necessary.
For more information on this, and how to apply and follow the GNU AGPL, see
<https://www.gnu.org/licenses/>.
*******************************************************************************
--]]

--wrapper logging function for this file
local function Log(...)
    Fizzle_FuzeLogMessage("HydrolysisReactor", ...)
end

DefineClass.HydrolysisReactor = {
    __parents = { "OutsideBuildingWithShifts", "WaterProducer", "AirProducer", "ElectricityProducer", "ResourceProducer", "DustGenerator", "BuildingDepositExploiterComponent",
                  "SubsurfaceDepositConstructionRevealer", },
    subsurface_deposit_class = "SubsurfaceDepositWater",

    building_update_time = const.HourDuration,
    track_multiple_hit_moments_in_work_state = true,

    stockpile_spots1 = { "Resourcepile" },
    additional_stockpile_params1 = {
        apply_to_grids = false,
        has_platform = true,
        snap_to_grid = false,
        priority = 2,
        additional_supply_flags = const.rfSpecialDemandPairing
    },

    anim_moments_thread = false,
    water_production = 2000,
    air_production = 3000,
    electricity_production = 0,
}

function HydrolysisReactor:GetEntityNameForPipeConnections(grid_skin_name)
    return grid_skin_name ~= "Default" and "Moxie" .. grid_skin_name or "Moxie"
end

function HydrolysisReactor:CreateElectricityElement()
    ElectricityProducer.CreateElectricityElement(self)
end

function HydrolysisReactor:CreateLifeSupportElements()
    AirProducer.CreateLifeSupportElements(self)
    WaterProducer.CreateLifeSupportElements(self)
end

function HydrolysisReactor:SetPriority(priority)
    OutsideBuildingWithShifts.SetPriority(self, priority)
    AirProducer.SetPriority(self, priority)
end

function HydrolysisReactor:ShouldShowNotConnectedToLifeSupportGridSign()
    return AirProducer.ShouldShowNotConnectedToLifeSupportGridSign(self)
end

function HydrolysisReactor:BuildingUpdate()
    if self.working then
        Log("Tick")
        self:ProduceSupply("water", self.air_production)
    end
    RebuildInfopanel(self)
end

function HydrolysisReactor:DroneLoadResource(drone, request, resource, amount)
    TaskRequester.DroneLoadResource(self, drone, request, resource, amount)

    if not self.working then
        self:UpdateWorking()
    end
end

function HydrolysisReactor:GameInit()
    self:DepositChanged()
end

function HydrolysisReactor:OnDepositsLoaded()
    self:DepositChanged()
    self:UpdateConsumption()
    self:UpdateWorking()
end

function HydrolysisReactor:ProduceSupply(resource, amount)
    if resource == "water" and self.nearby_deposits[1] then
        local deposit_grade = self.nearby_deposits[1].grade --ExtractResource may kill this

        if self:DoesHaveUpgradeConsumption() then
            amount = self:Consume_Upgrades_Production(amount, 100)
        end

        amount = self:ExtractResource(amount)
        if self:ProduceWasteRock(amount, deposit_grade) then
            self:UpdateWorking(false)
        end
    end
end

function HydrolysisReactor:DepositChanged()
    local deposit_multiplier = self:GetCurrentDepositQualityMultiplier()
    local amount = MulDivRound(self:GetClassValue("water_production"), deposit_multiplier, 100)
    self:SetBase("water_production", amount)
    self:UpdateWorking()
end

function HydrolysisReactor:OnChangeActiveDeposit()
    BuildingDepositExploiterComponent.OnChangeActiveDeposit(self)
    self:DepositChanged()
end

function HydrolysisReactor:OnDepositDepleted(deposit)
    BuildingDepositExploiterComponent.OnDepositDepleted(self, deposit)
    self:DepositChanged()
end

function HydrolysisReactor:UpdateElectricityProduction()
    --if type(self.upgrades_built) == "table" then -- to avoid error in self:HasUpgrade >=(
    if self:HasUpgrade("HydrolysisReactor_AdvancedReactions") and self:IsUpgradeOn("HydrolysisReactor_AdvancedReactions") then
        self:SetBase("electricity_production", 2 * self.air_production)
    else
        self:SetBase("electricity_production", 0)
    end
    --end
end

function HydrolysisReactor:OnSetWorking(working)
    Building.OnSetWorking(self, working)
    AirProducer.OnSetWorking(self, working)
    ElectricityProducer.OnSetWorking(self, working)

    local production = working and self.water_production or 0
    if self.water then
        self.water:SetProduction(production, production)
    end

    self:UpdateElectricityProduction()
end

function HydrolysisReactor:IsIdle()
    return self.ui_working and not self:CanExploit() and not self.city.colony:IsTechResearched("NanoRefinement")
end

function HydrolysisReactor:SetUIWorking(working)
    Building.SetUIWorking(self, working)
    BuildingDepositExploiterComponent.UpdateIdleExtractorNotification(self)
end

function HydrolysisReactor:UpdateAttachedSigns()
    if self.electricity then
        self:AttachSign(self:ShouldShowNotConnectedToPowerGridSign(), "SignNoPowerProducer")
    end
end

function HydrolysisReactor:OnUpgradeToggled()
    self:UpdateElectricityProduction()
end

function HydrolysisReactor:OnModifiableValueChanged(prop)
    Log("prop = ", prop)
    if prop == "water_production" and self.water then
        local production = self.working and self.water_production or 0
        self.water:SetProduction(production, production)
        Log("val = ", self.water_production)
    end
    if prop == "air_production" and self.air then
        self.air:SetProduction(self.working and self.air_production or 0)
        Log("val = ", self.air_production)

        self:UpdateElectricityProduction()
    end
    if (prop == "electricity_production" or prop == "performance") and self.electricity then
        Log("val = ", self.electricity_production)
        self.electricity:SetProduction(self.working and self:GetPerformanceModifiedElectricityProduction() or 0)
    end
end

Code/BuildingTemplate.lua
--Copyright
--[[
*******************************************************************************
Fizzle_Fuze's Surviving Mars Mods
Copyright (c) 2022 Fizzle Fuze Enterprises (mods@fizzlefuze.com)

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU Affero General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Affero General Public License for more details.

    You should have received a copy of the GNU Affero General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.

  If your software can interact with users remotely through a computer
network, you should also make sure that it provides a way for users to
get its source.  For example, if your program is a web application, its
interface could display a "Source" link that leads users to an archive
of the code.  There are many ways you could offer source, and different
solutions will be better for different programs; see section 13 for the
specific requirements.

  You should also get your employer (if you work as a programmer) or school,
if any, to sign a "copyright disclaimer" for the program, if necessary.
For more information on this, and how to apply and follow the GNU AGPL, see
<https://www.gnu.org/licenses/>.
*******************************************************************************
--]]

--wrapper logging function for this file
local function Log(...)
        Fizzle_FuzeLogMessage("BuildingTemplate", ...)
end

--translation strings
local Translate = { ID = {}, Text = {} }

Translate.Text['display_name'] = "Hydrolysis Reactor"
Translate.Text['display_name_pl'] = "Hydrolysis Reactors"
Translate.Text['description'] = "Extracts Oxygen from Water.<newline><newline>All extractors raise dust resulting in more frequent maintenance for buildings in the grey area."
Translate.Text['upgrade2_display_name'] = "Advanced Reactions"
Translate.Text['upgrade2_description'] = "Advanced Reactions allow excess energy from hydrogen atoms to be used in the power grid."
Translate.Text['upgrade3_display_name'] = "Moisture Farming"
Translate.Text['upgrade3_description'] = "Allows reactors to take moisture from the atmosphere to produce more oxygen and power."
Translate.Text['encyclopedia_text'] = "Uses the power of science to extract oxygen from water."

--get every string a unique ID
for k, _ in pairs(Translate.Text) do
        Translate.ID[k] = RandomLocId()
        if not Translate.ID[k] then
                Log("ERROR", "Could not find valid translation ID for '", k, "'!")
        end
end

function OnMsg.ClassesPostprocess()
        Log("Adding BT")
        if not BuildingTemplates.FFHydrolysisReactor then
                PlaceObj('BuildingTemplate', {
                        'comment', "Hydrolysis Reactor",
                        'Group', "Life-Support",
                        'Id', "FFHydrolysisReactor",
                        'template_class', "HydrolysisReactor",
                        'pin_rollover_context', "air",
                        'pin_rollover_hint', T(471070805886, --[[ModItemBuildingTemplate FF-Hydrolysis-Reactor pin_rollover_hint]] "<image UI/Infopanel/left_click.tga 1400> Select"),
                        'pin_rollover_hint_xbox', T(488123886351, --[[ModItemBuildingTemplate FF-Hydrolysis-Reactor pin_rollover_hint_xbox]] "<image UI/PS4/Cross.tga> View"),
                        'construction_cost_Concrete', 5000,
                        'construction_cost_Metals', 4000,
                        'construction_cost_Electronics', 1000,
                        'construction_cost_MachineParts', 2000,
                        'build_points', 1500,
                        'is_tall', true,
                        'dome_forbidden', false,
                        'upgrade1_id', "MOXIE_MagneticFiltering",
                        'upgrade1_display_name', T(396174688820, --[[ModItemBuildingTemplate FF-Hydrolysis-Reactor upgrade1_display_name]] "Magnetic Filtering"),
                        'upgrade1_description', T(936522085574, --[[ModItemBuildingTemplate FF-Hydrolysis-Reactor upgrade1_description]] "+<upgrade1_mul_value_1>% Production."),
                        'upgrade1_icon', "UI/Icons/Upgrades/amplify_01.tga",
                        'upgrade1_mod_label_1', "MOXIE",
                        'upgrade1_mod_prop_id_1', "air_production",
                        'upgrade1_mul_value_1', 50,
                        'upgrade1_upgrade_cost_Concrete', 2000,
                        'upgrade1_upgrade_cost_Polymers', 5000,
                        'upgrade2_id', "HydrolysisReactor_AdvancedReactions",
                        'upgrade2_display_name', T(Translate.ID['upgrade2_display_name'], Translate.Text['upgrade2_display_name']),
                        'upgrade2_description', T(Translate.ID['upgrade2_description'], Translate.Text['upgrade2_description']),
                        'upgrade2_icon', "UI/Icons/Upgrades/eternal_fusion_01.tga",
                        'upgrade2_mod_prop_id_1', "electricity_production",
                        'upgrade2_add_value_1', 0,
                        'upgrade2_upgrade_cost_Electronics', 2000,
                        'upgrade2_upgrade_cost_PreciousMetals', 1000,
                        'upgrade3_id', "HydrolysisReactor_MoistureFarming",
                        'upgrade3_display_name', T(Translate.ID['upgrade3_display_name'], Translate.Text['upgrade3_display_name']),
                        'upgrade3_description', T(Translate.ID['upgrade3_description'], Translate.Text['upgrade3_description']),
                        'upgrade3_icon', "UI/Icons/Upgrades/hygroscopic_coating_01.tga",
                        'upgrade3_mod_prop_id_1', "air_production",
                        'upgrade3_mul_value_1', 50,
                        'upgrade3_upgrade_cost_Metals', 2000,
                        'upgrade3_upgrade_cost_Polymers', 5000,
                        'maintenance_resource_type', "MachineParts",
                        'maintenance_resource_amount', 2000,
                        'maintenance_threshold_base', 150000,
                        'sight_category', "Additional Buildings",
                        'sight_satisfaction', 5,
                        'display_name', T(Translate.ID['display_name'], Translate.Text['display_name']),
                        'display_name_pl', T(Translate.ID['display_name_pl'], Translate.Text['display_name_pl']),
                        'description', T(Translate.ID['description'], Translate.Text['description']),
                        'build_category', "Life-Support",
                        'display_icon', CurrentModPath.."UI/Icons/electrolyzer.png",
                        'build_pos', 5,
                        'entity', "Moxie",
                        'show_range', true,
                        'encyclopedia_id', "HydrolysisReactor",
                        'encyclopedia_text', T(Translate.ID['encyclopedia_text'], Translate.Text['encyclopedia_text']),
                        'encyclopedia_image', "UI/Encyclopedia/Electrolyzer.tga",
                        'label1', "OutsideBuildings",
                        'label2', "OutsideBuildingsTargets",
                        'palette_color1', "outside_accent_1",
                        'palette_color2', "outside_metal",
                        'palette_color3', "life_base",
                        'demolish_sinking', range(1, 5),
                        'demolish_debris', 85,
                        'disabled_in_environment1', "Asteroid",
                        'air_production', 3000,
                        'water_production', 2000,
                        'exploitation_resource', "Water",
                })

        end
end